### PR TITLE
Add CROSS

### PR DESCRIPTION
**Pull request by rtjk**
Tracker for [liboqs#1881](https://github.com/open-quantum-safe/liboqs/pull/1881) which adds the CROSS signature scheme.

**NOTE:** The Object Identifiers in `generate.yml` are placeholders for now, we are in the process of obtaing new OIDs for CROSS.

### CODE REVIEW

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:1528]: Good choice of code point (counting on)!

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:1527]: Not so good choice of OID. We simply keep counting up the "free" OID space "1.3.9999.x.y" I'd suggest you simply use x=9 and count y up through the Cross variants (just as Mayo did with x=8). That alleviates us of having to go through some official OID-approval procedure.

Comment (rtjk) [reply oqs-template/generate.yml]: Thanks @baentsch, switched to OIDs from 1.3.9999.9.1 to 1.3.9999.9.18

File: oqs-template/generate.yml
Comment (rtjk) [reply oqs-template/generate.yml]: Undestood, only CROSS-rsdp-128-balanced is enbled now

Comment (rtjk) [reply oqs-template/generate.yml]: We just managed to get official OIDs from IANA, should we switch to those? @baentsch

File: oqs-template/generate.yml
Comment (rtjk) [reply oqs-template/generate.yml]: We just managed to get official OIDs from IANA, should we switch to those? @baentsch

Comment (rtjk) [reply oqs-template/generate.yml]: Ok OIDs for CROSS are now from 1.3.6.1.4.1.62245.2.1.1 to 1.3.6.1.4.1.62245.2.1.18 and I added myself as code owner for genrate.yml

Comment (baentsch) [reply oqs-template/generate.yml]: Thanks, @rtjk . Marking this resolved then. Would you want to re-base and/or investigate the CI failures first? Or would you want me to look into the failure reasons in CI?

**Comment by baentsch** [comment]
@SWilson4 Do you know how one can trigger GH CI on this PR again? Our CCI script clearly has a bug (indicating a pass while failing) but I'm not overly interested in fixing that given #248. Second question: How can one trigger a (new) CI run using the CROSS PR/branch in `liboqs` (and not "main")? Thanks in advance for suggestions/HowTos/pointers!

**Comment by baentsch** [comment]
> It looks to me like CI is failing because at various points liboqs' main branch is being cloned instead of my fork; the build process fails because it can't find CROSS, which is not part of main yet.

The `liboqs` branch referenced exactly can be set to a different value when triggering CI at `liboqs` for such downstream test. Normally, this requires a specific commit tag -- but I'm not sure how to trigger a re-run of [that PR](https://github.com/open-quantum-safe/liboqs/pull/1881). Hence [my question above](https://github.com/open-quantum-safe/oqs-provider/pull/461#issuecomment-2296303232) to @SWilson4 in this regard. @rtjk please note that https://github.com/open-quantum-safe/liboqs/pull/1881 also needs re-basing....

File: .github/CODEOWNERS
Comment (baentsch) [inline .github/CODEOWNERS:5]: I currently take this as your agreement to be a [committer](https://github.com/open-quantum-safe/oqs-provider/blob/main/GOVERNANCE.md#committers) in the project. Please review the link to see which obligations and permissions this (would) entail. If you do agree, please also add yourself to the corresponding README section. The addition to the GOVERNANCE.md file I'd suggest doing after some months of good cooperation: OK, @rtjk ? If you don't want that responsibility, no problem, but then the entry above needs to be removed to eliminate the error message. Please replace it by a comment with your email address in generate.yml for the Cross OIDs so we know whom to contact in case of problems.

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:1528]: Is it intentional that no hybrid forms ("mix_with") are specified?

**Review by baentsch — DISMISSED** [review]
Functionally and wrt testing LGTM. See single comments regarding "management & support".

**Comment by SWilson4** [comment]
> @SWilson4 Do you know how one can trigger GH CI on this PR again? Our CCI script clearly has a bug (indicating a pass while failing) but I'm not overly interested in fixing that given #248. Second question: How can one trigger a (new) CI run using the CROSS PR/branch in `liboqs` (and not "main")? Thanks in advance for suggestions/HowTos/pointers!

For the first, you can go over to one of the jobs and hit "Re-run jobs" -> "Re-run all jobs". For the second, I wrote the '-tracker' script trigger code with release testing in mind, so I didn't take forks into account. I'm pretty sure it will only work for `open-quantum-safe` branches. :(

**Comment by baentsch** [comment]
> I'm pretty sure it will only work for open-quantum-safe branches. :(

That should be easy to establish: Just push the "add-cross[-tracker]" branches into the OQS name space and add a [trigger-downstream] commit... Only "little issue": It worked fine for `oqs-provider` but did not for `liboqs`: Is it possibly a GH permissions issue again? @ryjones : Should I be able to execute successfully the command `git push --set-upstream origin add_cross` in `liboqs`?

See status and command sequence below (& sorry for the German of my system's installation):

```
baentsch:~/git/oqs/liboqs$ git status
Auf Branch add-cross
Ihr Branch ist auf demselben Stand wie 'rtjk/add-cross'.

nichts zu committen, Arbeitsverzeichnis unverändert
baentsch:~/git/oqs/liboqs$ git push --set-upstream origin add_cross
error: Src-Refspec add_cross entspricht keiner Referenz.
error: Fehler beim Versenden einiger Referenzen nach 'github.com:open-quantum-safe/liboqs.git'

```

**Comment by SWilson4** [comment]
I just added an `add-cross` branch in `liboqs` and triggered the [release tests](https://github.com/open-quantum-safe/oqs-provider/actions/runs/10495257860/job/29073311761).

**Comment by SWilson4** [comment]
> ```
> baentsch:~/git/oqs/liboqs$ git status
> Auf Branch add-cross
> Ihr Branch ist auf demselben Stand wie 'rtjk/add-cross'.
> 
> nichts zu committen, Arbeitsverzeichnis unverändert
> baentsch:~/git/oqs/liboqs$ git push --set-upstream origin add_cross
> error: Src-Refspec add_cross entspricht keiner Referenz.
> error: Fehler beim Versenden einiger Referenzen nach 'github.com:open-quantum-safe/liboqs.git'
> ```

I could be wrong, but I think this was failing because the branch name uses a hyphen and the push command uses an underscore :/

**Comment by rtjk** [comment]
Looking at the [release test](https://github.com/open-quantum-safe/oqs-provider/actions/runs/10495257860/job/29073311761) I notice it enables all algorithms, so all 18 variants of CROSS are on.
https://github.com/open-quantum-safe/oqs-provider/blob/a0f475daba6dcbe3088ff8943810d15cecd2c971/scripts/release-test-ci.sh#L22-L23
And it fails to do a TLS handshake with one parameter set only: CROSSrsdp256fast.
```
SSL_accept() failed returning -1, SSL error 1.
809B9FB4A07F0000:error:0A0C0103:SSL routines:tls_construct_cert_verify:internal error:ssl/statem/statem_lib.c:415:
SSL_connect() failed returning -1, SSL error 1.
809B9FB4A07F0000:error:0A000438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error:ssl/record/rec_layer_s3.c:908:SSL alert number 80
  TLS-SIG handshake test failed: CROSSrsdp256fast, return code: -5
```
My first guess would be that the signature produced by this variant is too large, it looks like it's the only one in liboqs with `CRYPTO_BYTES` larger than 64K.

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:1575]: I am afraid you have to remove this algorithm here, @rtjk : As you assume, this has to do with its signature being longer than 64kB (I happen to know this limit as I instigated in OpenSSL the change from 16kB to the maximum permitted by spec at https://github.com/openssl/openssl/blob/e91384d5b0547bf797e2b44976f142d146c4e650/ssl/statem/statem_local.h#L34-L35 to allow more PQ algs to operate in OpenSSL). Thanks to @SWilson4 for triggering this CI!

Comment (baentsch) [reply oqs-template/generate.yml:1575]: And in case you want to check the spec: It's section B.3.3 in https://www.rfc-editor.org/rfc/rfc8446 specifying maximum signature length in TLS.

Comment (rtjk) [reply oqs-template/generate.yml]: Got it, I just removed rsdp-256-fast and shifted the code-points accordingly.

Comment (baentsch) [reply oqs-template/generate.yml]: Thanks. Triggered a full re-test: https://github.com/open-quantum-safe/oqs-provider/actions/runs/10506136727

Comment (baentsch) [reply oqs-template/generate.yml]: But, @rtjk it seems you'd need to re-run [code formatting](https://github.com/open-quantum-safe/oqs-provider/blob/main/CONTRIBUTING.md#coding-style).

Comment (rtjk) [reply oqs-template/generate.yml]: Just done :)

File: .github/CODEOWNERS
Comment (baentsch) [reply .github/CODEOWNERS:5]: As per https://github.com/open-quantum-safe/liboqs/pull/1881#discussion_r1726991735 suggest to not add yourself to the CODEOWNERS file. Please add your (or your advisors') github handles by comment next to the generate.yml entries for Cross instead.

**Comment by baentsch** [comment]
@rtjk FYI, the full "release-level" test got _much_ farther now, but then still failed: See [log](https://github.com/open-quantum-safe/oqs-provider/actions/runs/10508804690/job/29113504897). You should be able to locally reproduce by running `./scripts/release-test.sh`.

File: oqs-template/generate.yml
Comment (rtjk) [reply oqs-template/generate.yml:1528]: Would it be ok to address hybrids in a future PR? I'd like to make sure I fully understand how they are composed first.

File: .github/CODEOWNERS
Comment (rtjk) [reply .github/CODEOWNERS]: @alexrow has [agreed](https://github.com/open-quantum-safe/liboqs/pull/1881#discussion_r1732046494) to be a committer! Holding off on the GOVERNANCE file as you suggested @baentsch.

File: oqs-template/generate.yml
Comment (baentsch) [reply oqs-template/generate.yml:1528]: Of course!

File: .github/CODEOWNERS
Comment (ryjones) [inline .github/CODEOWNERS:5]: we'll need an update to the [governance yaml](https://github.com/open-quantum-safe/tsc/blob/main/config.yaml) to support this user

**Comment by SWilson4** [comment]
The proposed configuration update in https://github.com/open-quantum-safe/tsc/pull/77 will fix the CODEOWNERS file issue by providing @alexrow with write permissions (which is necessary for the CODEOWNERS file to work as intended).

**Comment by praveksharma** [comment]
Closing since #530 has landed.

**Comment by baentsch** [comment]
@rtjk @alexrow Please accept my sincere apologies for closing and replacing your PR with #530: While I have not been party to this approach nor approve of it as maintainer valuing every contributor's work with an interest to see it come to completion I think it has been borne out of interest by @praveksharma and @SWilson4 to make progress and move `oqsprovider` towards release and alleviating you of tedious work (rebasing). @alexrow please see #531 with an action on you.

**Comment by rtjk** [comment]
@baentsch @praveksharma No worries at all, I totally understand the situation. Happy to see CROSS merged in oqs-provider too :)

**Pull request closed without merging**
